File: leetcode-js/138-copy-list-with-random-pointer.js
/**
 * Definition for singly-linked list with a random pointer.
 * function RandomListNode(label) {
 *     this.label = label;
 *     this.next = this.random = null;
 * }
 */

/**
 * @param {RandomListNode} head
 * @return {RandomListNode}
 */
var copyRandomList = function(head) {
    if (!head) {
        return null;
    }

    // 1. Insert copied nodes

    let node = head;

    while (node) {
        const copy = new RandomListNode(node.label);
        copy.next = node.next;
        node.next = copy;
        node = copy.next;
    }

    // 2. Copy random pointer

    node = head;

    while (node) {
        if (node.random) {
            node.next.random = node.random.next;
        }
        node = node.next.next;
    }

    // 3. Decouple linked lists

    node = head;
    let copiedNode = node.next;
    const copiedHead = node.next;

    while (node) {
        node.next = copiedNode.next;
        node = node.next;
        if (copiedNode.next) {
            copiedNode.next = copiedNode.next.next;
        }
        copiedNode = copiedNode.next;
    }

    return copiedHead;
};

File: leetcode-java/138-copy-list-with-random-pointer.java
/**
 * Definition for singly-linked list with a random pointer.
 * class RandomListNode {
 *     int label;
 *     RandomListNode next, random;
 *     RandomListNode(int x) { this.label = x; }
 * };
 */
public class Solution {
    public RandomListNode copyRandomList(RandomListNode head) {
        if (head == null) {
            return null;
        }

        RandomListNode node = head;

        while (node != null) {
            RandomListNode copy = new RandomListNode(node.label);
            copy.next = node.next;
            node.next = copy;
            node = node.next.next;
        }

        node = head;

        while (node != null) {
            if (node.random != null) {
                node.next.random = node.random.next;
            }
            node = node.next.next;
        }

        node = head;
        RandomListNode copyHead = node.next;
        RandomListNode copyNode = copyHead;

        while (node != null) {
            node.next = node.next.next;
            node = node.next;
            if (copyNode.next != null) {
                copyNode.next = copyNode.next.next;
            }
            copyNode = copyNode.next;
        }

        return copyHead;
    }
}
